File: JAVA/Recamans sequence.java
class Solution{
    static ArrayList<Integer> recamanSequence(int n){
        ArrayList<Integer> al = new ArrayList<>();
        HashSet<Integer> h = new HashSet<>();
        al.add(0);
        h.add(0);
        for (int i=1;i<n;i++){
            if(al.get(i-1)-i>0 && h.contains(al.get(i-1)-i)==false){
                al.add(al.get(i-1)-i);
                h.add(al.get(i-1)-i);
            }else{
                al.add(al.get(i-1)+i);
                h.add(al.get(i-1)+i);
            }
        }
        return al;
    }
}

File: C++/Recamans sequence.cpp
class Solution{
public:
    vector<int> recamanSequence(int n){
        unordered_map<int,int> mp;
        vector<int> res(n);
        res[0] = 0;
        res[1] = 1;
        mp[0] = 0;
        mp[1] = 1;
        for(int i =2; i<n; i++)
        {
            int x = res[i-1] -i;
            if(x > 0 && mp[x] == 0)
            {
                res[i] = x;
                mp[x]++;
            }
            else
            {
                x = res[i-1] + i;
                res[i] = x;
                mp[x]++;
            }
        }
        return res;
    }
};
